MaxText/layers/attentions.py

```diff
@@ -260,22 +241,174 @@ def key_rotary(self, key: Array, inputs_positions: Array):
       name='key_rotary')(inputs=key, position=inputs_positions)
     return key
 
-  def apply_attention(
+
+  # Following Pallas MHA Flash Attention Reference.
+  # https://github.com/google/jax/blob/main/jax/experimental/pallas/ops/tpu/flash_attention.py
+  def generate_attention_mask(
       self,
+      query,
+      key,
+      decoder_segment_ids: Array | None,
+      decode: bool = False,
+  ) -> Array | None:
+    mask = None
+    if decoder_segment_ids is not None:
+      mask = decoder_segment_ids[:, :, None] == decoder_segment_ids[:, None, :]
+      mask = mask[:, None, None,:, :]
+
+    _, q_seq_len, _, _ = query.shape
+    _, kv_seq_len, _, _ = key.shape
+    mask_shape = (q_seq_len, kv_seq_len)
+    row_ids = jax.lax.broadcasted_iota(jnp.int32, mask_shape, 0)
+    col_ids = jax.lax.broadcasted_iota(jnp.int32, mask_shape, 1)
+    causal_mask = (col_ids <= row_ids)[None, None, None, :, :]
+
+    mask = causal_mask if mask is None else jnp.logical_and(mask, causal_mask)
+    return jnp.where(mask, 0.0, DEFAULT_MASK_VALUE) if mask is not None else None
+  
+  def apply_attention(self,
       query: Array,
       key: Array,
       value: Array,
       decoder_segment_ids: Array | None,
-      attention_bias: Array | None,
       dropout_rng: PRNGKey | None,
       deterministic: bool,
-      decode: bool = False,
+      decode: bool = False) -> Array:
+    
+    if self.attention_kernel == "dot_product":
```

Comment:
Let's consider to make attention_kernel as sub class and not to avoid this if-else branches.

Reply:
I'm actually comfortable with the `if-else` branching. MaxText is a much smaller project than say Pax so it is less important to make it possible for devs to "inject" new attention.

Reply:
OK then.